File: real_estate_be/internal/dto/deposit.go
package dto

// ── Request ──────────────────────────────────────────────

// CreateDepositRequest — khách điền form đặt cọc xem nhà.
// KHÔNG nhận amount/broker_fee từ client: hệ thống tự tra chính sách theo giá BĐS
// để tránh khách đặt cọc số tiền vô nghĩa làm mất tác dụng chống bùng.
type CreateDepositRequest struct {
	RealEstateID  uint64 `json:"real_estate_id"`
	ViewingDate   string `json:"viewing_date"`  // "2006-01-02"
	ViewingStart  string `json:"viewing_start"` // "HH:MM"
	ViewingEnd    string `json:"viewing_end"`   // "HH:MM"
	PaymentMethod string `json:"payment_method"` // VNPAY / MOMO / ZALOPAY
	BankCode      string `json:"bank_code"`
}

// BookingOptionsResponse — mức cọc + phí môi giới hệ thống đề xuất cho 1 BĐS.
// FE chỉ hiển thị, khách không sửa được.
type BookingOptionsResponse struct {
	RealEstateID    uint64  `json:"real_estate_id"`
	RealEstateTitle string  `json:"real_estate_title"`
	PriceVND        float64 `json:"price_vnd"`
	Amount          float64 `json:"amount"`
	BrokerFee       float64 `json:"broker_fee"`
	// Nhãn phân khúc giá đang áp dụng, VD "Từ 3 đến 5 tỷ"
	PolicyLabel string `json:"policy_label"`
	// true khi không khớp chính sách nào và phải dùng mức mặc định trong config
	IsFallback bool `json:"is_fallback"`
}

// RejectDepositRequest — môi giới từ chối lịch kèm lý do
type RejectDepositRequest struct {
	Reason string `json:"reason"`
}

// CheckinRequest — khách nhập OTP do môi giới hiển thị tại chỗ
type CheckinRequest struct {
	OTP string `json:"otp"`
}

// ReportResultRequest — 1 bên báo cáo kết quả buổi xem.
// Giai đoạn đã check-in (báo mua/không mua) BẮT BUỘC kèm bằng chứng ảnh.
// Riêng khách khai BOUGHT còn phải chọn loại tài liệu mua bán (PurchaseProof).
type ReportResultRequest struct {
	Report       string   `json:"report"`
	EvidenceURLs []string `json:"evidence_urls"`
	// Chỉ dùng khi khách khai BOUGHT: PURCHASE_CONTRACT / DEPOSIT_SLIP / PAYMENT_SLIP
	PurchaseProof string `json:"purchase_proof"`
}

// CreateDisputeRequest — một bên mở tranh chấp
type CreateDisputeRequest struct {
	Reason       string   `json:"reason"`
	EvidenceURLs []string `json:"evidence_urls"`
}

// AddEvidenceRequest — upload thêm bằng chứng vào dispute đang mở
type AddEvidenceRequest struct {
	EvidenceURLs []string `json:"evidence_urls"`
}

// ResolveDisputeRequest — admin ra quyết định xử lý tiền
type ResolveDisputeRequest struct {
	Resolution string `json:"resolution"` // REFUND_CUSTOMER / TRANSFER_BROKER / SPLIT
	Note       string `json:"note"`
	// Chỉ dùng khi resolution = SPLIT: % tiền cọc hoàn cho khách (0-100)
	SplitCustomerPercent int `json:"split_customer_percent"`
}

// PurchaseDecisionRequest — admin duyệt hoặc từ chối tài liệu mua nhà.
// Chỉ khi duyệt thì đơn mới sang VISITED_BOUGHT và tồn kho dự án mới bị trừ 1 căn.
type PurchaseDecisionRequest struct {
	Approved bool   `json:"approved"`
	Note     string `json:"note"`
}

// RateBrokerRequest — khách đánh giá môi giới sau buổi xem
type RateBrokerRequest struct {
	Rating  int    `json:"rating"`
	Comment string `json:"comment"`
}

// ── Response ─────────────────────────────────────────────

// DepositResponse — dữ liệu 1 deposit trả về cho FE (đã làm phẳng, không lộ otp_hash)
type DepositResponse struct {
	ID           uint64 `json:"id"`
	Status       string `json:"status"`
	CustomerID   uint64 `json:"customer_id"`
	CustomerName string `json:"customer_name"`
	CustomerPhone string `json:"customer_phone"`
	CustomerEmail string `json:"customer_email"`
	BrokerID     uint64 `json:"broker_id"`
	BrokerName   string `json:"broker_name"`
	BrokerPhone  string `json:"broker_phone"`
	BrokerEmail  string `json:"broker_email"`

	RealEstateID      uint64  `json:"real_estate_id"`
	RealEstateTitle   string  `json:"real_estate_title"`
	RealEstateAddress string  `json:"real_estate_address"`
	RealEstateSlug    string  `json:"real_estate_slug"`
	RealEstateThumb   string  `json:"real_estate_thumbnail"`
	ProjectID         *uint64 `json:"project_id"`

	Amount        float64  `json:"amount"`
	BrokerFee     float64  `json:"broker_fee"`
	RefundAmount  *float64 `json:"refund_amount"`
	PenaltyAmount *float64 `json:"penalty_amount"`

	ViewingDate  string `json:"viewing_date"`
	ViewingStart string `json:"viewing_start"`
	ViewingEnd   string `json:"viewing_end"`

	PaymentMethod string `json:"payment_method"`
	PaymentRef    string `json:"payment_ref"`
	PaidAt        string `json:"paid_at"`

	BrokerCheckin      bool   `json:"broker_checkin"`
	CustomerCheckin    bool   `json:"customer_checkin"`
	BrokerReport       string `json:"broker_report"`
	CustomerReport     string `json:"customer_report"`
	// Bằng chứng kèm báo cáo mua/không mua của từng bên
	BrokerReportEvidence   []string `json:"broker_report_evidence"`
	CustomerReportEvidence []string `json:"customer_report_evidence"`
	// Loại tài liệu khách xuất trình khi khai đã mua nhà
	CustomerPurchaseProof string `json:"customer_purchase_proof"`
	BrokerConfirmedAt      string   `json:"broker_confirmed_at"`
	RejectReason           string   `json:"reject_reason"`
	ReportDeadline         string   `json:"report_deadline"`

	// Cờ tiện dụng cho FE
	CanConfirm bool `json:"can_confirm"` // môi giới còn được xác nhận/từ chối
	CanCheckin bool `json:"can_checkin"` // được phép check-in OTP
	CanReport  bool `json:"can_report"`  // được phép báo cáo kết quả
	// Admin còn phải duyệt tài liệu mua nhà của đơn này
	CanApprovePurchase bool `json:"can_approve_purchase"`
	HasDispute         bool `json:"has_dispute"`
	HasRating          bool `json:"has_rating"`

	// Kết quả admin duyệt tài liệu mua nhà
	PurchaseDecisionBy   *uint64 `json:"purchase_decision_by"`
	PurchaseDecisionAt   string  `json:"purchase_decision_at"`
	PurchaseDecisionNote string  `json:"purchase_decision_note"`

	Dispute *DisputeResponse `json:"dispute,omitempty"`
	Rating  *BrokerRatingResponse `json:"rating,omitempty"`

	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

// WithdrawalBreakdown — chi tiết dòng tiền của 1 deposit (hiển thị cho admin)
type TransactionResponse struct {
	ID        uint64  `json:"id"`
	Type      string  `json:"type"`
	Amount    float64 `json:"amount"`
	Note      string  `json:"note"`
	CreatedAt string  `json:"created_at"`
}

type DisputeResponse struct {
	ID               uint64   `json:"id"`
	DepositID        uint64   `json:"deposit_id"`
	RaisedBy         string   `json:"raised_by"`
	Reason           string   `json:"reason"`
	EvidenceURLs     []string `json:"evidence_urls"`
	Status           string   `json:"status"`
	Resolution       string   `json:"resolution"`
	ResolvedBy       *uint64  `json:"resolved_by"`
	ResolvedAt       string   `json:"resolved_at"`
	EvidenceDeadline string   `json:"evidence_deadline"`
	CreatedAt        string   `json:"created_at"`
	// Thông tin deposit kèm theo cho màn admin
	Deposit *DepositResponse `json:"deposit,omitempty"`
}

type BrokerRatingResponse struct {
	ID         uint64 `json:"id"`
	DepositID  uint64 `json:"deposit_id"`
	BrokerID   uint64 `json:"broker_id"`
	CustomerID uint64 `json:"customer_id"`
	Rating     int    `json:"rating"`
	Comment    string `json:"comment"`
	CreatedAt  string `json:"created_at"`
}

// EscrowSummaryResponse — số liệu tổng quan quỹ escrow cho admin
type EscrowSummaryResponse struct {
	// Tổng tiền đang giữ (tiền cọc đã thanh toán, chưa release)
	HoldingAmount float64 `json:"holding_amount"`
	// Tổng đã hoàn cho khách
	RefundedAmount float64 `json:"refunded_amount"`
	// Tổng đã chuyển cho môi giới
	TransferredAmount float64 `json:"transferred_amount"`
	// Tổng tiền phạt thu từ bảo lãnh môi giới
	PenaltyAmount float64 `json:"penalty_amount"`
	// Số deposit theo từng trạng thái
	StatusCounts map[string]int64 `json:"status_counts"`
	// Số dispute đang mở
	OpenDisputes int64 `json:"open_disputes"`
}

// PaymentCallbackResponse — kết quả xác nhận thanh toán trả cho FE
type PaymentCallbackResponse struct {
	DepositID uint64 `json:"deposit_id"`
	Status    string `json:"status"`
	Success   bool   `json:"success"`
	Message   string `json:"message"`
}

// CreateDepositResponse — kết quả tạo đặt cọc: kèm URL để redirect khách đi thanh toán
type CreateDepositResponse struct {
	Deposit    DepositResponse `json:"deposit"`
	PaymentURL string          `json:"payment_url"`
	Gateway    string          `json:"gateway"`
	IsMock     bool            `json:"is_mock"`
}

// CheckinOTPResponse — OTP check-in hiển thị cho môi giới tại chỗ
type CheckinOTPResponse struct {
	OTP       string `json:"otp"`
	ExpiresAt string `json:"expires_at"`
}

File: real_estate_fe/app/types/deposit.ts
/**
 * Kiểu dữ liệu luồng đặt cọc escrow (khớp DTO backend internal/dto/deposit.go).
 */

export type DepositStatus =
  | 'AWAITING_PAYMENT'
  | 'PENDING'
  | 'BROKER_REJECTED'
  | 'BROKER_CONFIRMED'
  | 'CHECKED_IN'
  | 'PENDING_PURCHASE_APPROVAL'
  | 'VISITED_BOUGHT'
  | 'VISITED_NOT_BUY'
  | 'NO_SHOW_CUSTOMER'
  | 'NO_SHOW_BROKER'
  | 'DISPUTE'
  | 'REFUNDED'
  | 'COMPLETED'
  | 'CANCELLED'

export type PaymentMethod = 'VNPAY' | 'MOMO' | 'ZALOPAY'

export type DisputeResolution = 'REFUND_CUSTOMER' | 'TRANSFER_BROKER' | 'SPLIT'

/** Loại tài liệu khách phải xuất trình khi khai đã mua nhà */
export type PurchaseProof = 'PURCHASE_CONTRACT' | 'DEPOSIT_SLIP' | 'PAYMENT_SLIP'

export const PURCHASE_PROOF_OPTIONS: { label: string; value: PurchaseProof; hint: string }[] = [
  { label: 'Hợp đồng mua bán', value: 'PURCHASE_CONTRACT', hint: 'Hợp đồng mua bán đã ký giữa 2 bên' },
  { label: 'Phiếu đặt cọc mua nhà', value: 'DEPOSIT_SLIP', hint: 'Phiếu đặt cọc mua nhà có chữ ký' },
  { label: 'Biên nhận chuyển tiền', value: 'PAYMENT_SLIP', hint: 'Biên nhận/ủy nhiệm chi đã chuyển tiền mua' },
]

export const PURCHASE_PROOF_LABEL: Record<string, string> = {
  PURCHASE_CONTRACT: 'Hợp đồng mua bán',
  DEPOSIT_SLIP: 'Phiếu đặt cọc mua nhà',
  PAYMENT_SLIP: 'Biên nhận chuyển tiền',
}

/** Vai trò của người đang xem đơn đặt cọc */
export type DepositActorRole = 'CUSTOMER' | 'BROKER' | 'ADMIN'

export interface DepositDispute {
  id: number
  deposit_id: number
  raised_by: 'CUSTOMER' | 'BROKER' | 'SYSTEM'
  reason: string
  evidence_urls: string[]
  status: 'OPEN' | 'REVIEWING' | 'RESOLVED'
  resolution: string
  resolved_by: number | null
  resolved_at: string
  evidence_deadline: string
  created_at: string
  deposit?: Deposit
}

export interface DepositRating {
  id: number
  deposit_id: number
  broker_id: number
  customer_id: number
  rating: number
  comment: string
  created_at: string
}

export interface Deposit {
  id: number
  status: DepositStatus
  customer_id: number
  customer_name: string
  customer_phone: string
  customer_email: string
  broker_id: number
  broker_name: string
  broker_phone: string
  broker_email: string

  real_estate_id: number
  real_estate_title: string
  real_estate_address: string
  real_estate_slug: string
  real_estate_thumbnail: string
  project_id: number | null

  amount: number
  broker_fee: number
  refund_amount: number | null
  penalty_amount: number | null

  viewing_date: string
  viewing_start: string
  viewing_end: string

  payment_method: PaymentMethod
  payment_ref: string
  paid_at: string

  broker_checkin: boolean
  customer_checkin: boolean
  broker_report: string
  customer_report: string
  /** Bằng chứng ảnh kèm báo cáo mua/không mua của từng bên */
  broker_report_evidence: string[]
  customer_report_evidence: string[]
  /** Loại tài liệu khách xuất trình khi khai đã mua nhà */
  customer_purchase_proof: string
  broker_confirmed_at: string
  reject_reason: string
  report_deadline: string

  can_confirm: boolean
  can_checkin: boolean
  can_report: boolean
  /** Admin còn phải duyệt tài liệu mua nhà của đơn này */
  can_approve_purchase: boolean
  has_dispute: boolean
  has_rating: boolean

  /** Kết quả admin duyệt tài liệu mua nhà */
  purchase_decision_by: number | null
  purchase_decision_at: string
  purchase_decision_note: string

  dispute?: DepositDispute
  rating?: DepositRating

  created_at: string
  updated_at: string
}

export interface CreateDepositPayload {
  real_estate_id: number
  viewing_date: string
  viewing_start: string
  viewing_end: string
  payment_method: PaymentMethod
}

/** Mức cọc + phí môi giới hệ thống đề xuất theo giá BĐS (khách không sửa được) */
export interface BookingOptions {
  real_estate_id: number
  real_estate_title: string
  price_vnd: number
  amount: number
  broker_fee: number
  /** Nhãn phân khúc giá đang áp dụng, VD "Từ 3 đến 5 tỷ" */
  policy_label: string
  /** true khi BĐS chưa có giá nên phải dùng mức mặc định */
  is_fallback: boolean
}

export interface CreateDepositResult {
  deposit: Deposit
  payment_url: string
  gateway: string
  is_mock: boolean
}

export interface CheckinOtpResult {
  otp: string
  expires_at: string
}

export interface PaymentCallbackResult {
  deposit_id: number
  status: DepositStatus
  success: boolean
  message: string
}

export interface EscrowSummary {
  holding_amount: number
  refunded_amount: number
  transferred_amount: number
  penalty_amount: number
  status_counts: Record<string, number>
  open_disputes: number
}

// Nhãn + màu tag Naive UI cho từng trạng thái đặt cọc
export const DEPOSIT_STATUS_META: Record<
  DepositStatus,
  { label: string; type: 'default' | 'info' | 'success' | 'warning' | 'error' }
> = {
  AWAITING_PAYMENT: { label: 'Chờ thanh toán', type: 'warning' },
  PENDING: { label: 'Chờ môi giới xác nhận', type: 'info' },
  BROKER_REJECTED: { label: 'Môi giới từ chối', type: 'error' },
  BROKER_CONFIRMED: { label: 'Đã xác nhận lịch', type: 'info' },
  CHECKED_IN: { label: 'Đã check-in', type: 'info' },
  PENDING_PURCHASE_APPROVAL: { label: 'Chờ duyệt tài liệu mua', type: 'warning' },
  VISITED_BOUGHT: { label: 'Đã mua nhà', type: 'success' },
  VISITED_NOT_BUY: { label: 'Đã xem - không mua', type: 'success' },
  NO_SHOW_CUSTOMER: { label: 'Khách không đến', type: 'error' },
  NO_SHOW_BROKER: { label: 'Môi giới không đến', type: 'error' },
  DISPUTE: { label: 'Đang tranh chấp', type: 'warning' },
  REFUNDED: { label: 'Đã hoàn tiền', type: 'success' },
  COMPLETED: { label: 'Đã tất toán', type: 'success' },
  CANCELLED: { label: 'Đã huỷ', type: 'default' },
}

export const PAYMENT_METHOD_LABEL: Record<PaymentMethod, string> = {
  VNPAY: 'VNPay',
  MOMO: 'Momo',
  ZALOPAY: 'ZaloPay',
}
